openDS term data schema

sharifX opened the pull request:
A JSON schema to describe the openDS terms as FDO.
The first version of this schema is focused on creating the basic structure of the object. Feedback from [RFC](https://docs.google.com/document/d/1m6Dzz2fOy3ZCLL9De8gHKH0Pw3TWY3Az-9pSyDAxbyY/edit) will be incorporated in different stages. There are a few decisions regarding RDF, JSON-LD we have to make.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+        "https://schema.org/Thing"
+      ]
+    },
+    "ods:repeatable": {
```
This and `ods:isRequired` feel like they should come from a generic ontology.
Maybe we could have a look at something like `owl:cardinality`

sharifX replied:
ok. yes, I was struggling to find one for this. owl:cardinality seems like a good option.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+        "https://schema.org/Thing"
+      ]
+    },
+    "ods:repeatable": {
```
This doesn't comply to our `ods` naming convention.
For booleans this is `ods:isXxx`
So in this case it should be `ods:isRepeatable` in line with `ods:isRequired`

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+        true
+      ]
+    },
+    "ods:isRequired": {
```
These might also be tricky when we switch belongToClass to an array. It could then be required for one class and be optional for another.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+      "type": "string",
+      "format": "uri",
+      "description": "License URI",
+      "enum": [
```
Can't we make this a constant. All terms minted by openDS (us) would get the same license right? Or might there be different licenses used for different terms?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+      ]
+    },
+   
+    "dcterms:licence": {
```
Mixing schema and dcterms. I would say stick with one or the other but mixing might be confusing. Most look to be coming from schema then I would suggest https://schema.org/license

samleeflang reviewed (commented):
Ok I think the start is there but I would make some changes:
- Don't mix schema.org and dcterms terms, there is a lot of overlap but I would say we pick one and stick with it
- Keep to naming convention
- BelongToClass probably needs to change but that impacts some other fields as well
-

wouteraddink commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://schemas.dissco.tech/schemas/fdo-type/term/latest/term.json",
```
I agree

wouteraddink commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+      "description": "The DOI to the FDO type of the object",
+      "pattern": "^https://doi\\.org/[\\w.]+/[\\w.]+",
+      "examples": [
+        "https://doi.org/10.15468/1a2b3c"
```
I think the FDO type should get a Handle not a DOI?
 regarding DOIs, I think I have seen a slightly different pattern used in code from Sou or Tom, including 10. This pattern is better, maybe we should use the same regex in all code. Note that the 21. examples are Handles not DOIs and may at some point not resolve through doi.org. Note also that DOIs expressed as DOI:10.1234/1234 are also valid DOIs.

sharifX replied:
the "21." is coming from the epIC DTR.

sharifX replied:
kept the example same now as the other digital objects.

wouteraddink commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+        "https://creativecommons.org/licenses/by/4.0/"
+      ]
+    },
+    "ods:belongsToClass": {
```
I would use the spdx urls for the machine readable version of the licenses, as we do in the FDO profiles also.

sharifX replied:
I got the "belongsToClass" idea from ltc. they only have one term but yes, we can make this an array.

sharifX commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+        }
+      ]
+    },
+     "schema:contributor": {
```
ok. yes. the definition mentions "secondary contributor".

sharifX replied:
removed contributor for now.

sharifX commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+        "https://hdl.handle.net/20.5000.1025/XXX-XXX-XXX"
+      ]
+    },
+    "ods:type": {
```
updated. in the new schema

sharifX commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,195 @@
+{
```
this in the json schema?

samleeflang replied:
Yes certainly, every term has examples in the schema

sharifX replied:
ok. will add this.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+        "ods:Term"
+      ]
+    },
+    "rdf:type": {
```
Isn't the `@type` the same as `rdf:type` ? But here they have different values, how would this work?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+        "ods:Active"
+      ]
+    },
+    "ods:version": {
```
Do we version the terms itself, at the moment we only version the DigitalObject. We could probably do this but don't we need more provenance attributes than? Such as `owl:versionInfo` or `owl:previousVersion` `skos:historyNote`, `dcterms:isVersionOf`, etc...?

sharifX replied:
check tdwg and add previous version

sharifX replied:
add Notes (for decision context)

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+    "dcterms:creator": {
+      "type": "object",
+      "description": "Contains information about the creator/proposer/requester of the term. This is an ods:Agent object",
+      "$ref": "https://schemas.dissco.tech/schemas/fdo-type/shared-model/0.4.0/agent.json",
```
I am not sure about the use of a openDS object here. The example would look a bit differently as the orcid should also be there as an identifier.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+    },
+    "dcterms:creator": {
+      "type": "object",
+      "description": "Contains information about the creator/proposer/requester of the term. This is an ods:Agent object",
```
These are three different roles but there is just room for a single creator. Can it occur that there are multiple creators or both a requester and a creator?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+      "const": "https://spdx.org/licenses/CC0-1.0.json",
+      "description": "The licence for the term, required to be always public domain"
+    },
+    "ods:belongsToClass": {
```
Wondering if `rdfs:range` does not cover this? This doesn't feel like we should be the ones creating this term

sharifX replied:
we are using rdfs:range. updated.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+        ["https://terms.dissco.tech/digital-specimen-terms#EntityRelationship", "http://purl.org/dc/dcmitype/Collection"]
+      ]
+    },
+    "owl:cardinality": {
```
Do we need all three, if we have the other two that would be enough right? Is there a use case when those two won't suffice?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,168 @@
+    }
+  },
+  "required": [
+    "@type",
```
Still missing `@id` here and I think more terms can/should be mandatory

samleeflang reviewed (commented):
Might be best to present this in the exploration meeting, I have difficulty grasping how this would function in the landscape? What will the identifiers look like, do we need changes to the terms (for example add version)? How does this fit with semantic best practise (we use owl terms but don't implement owl)?

sharifX commented:
updated schema 
changes in this version: 
- updated Regex for the Handle 
- removed rdf:Type 
- used ods:hasAgent instead of dcterms:creator 
- added skos:note and skos:example 
- removed owl cardinality. we use json schema required for this 
-  added hasVersion for more provenance metadata 
- used rdfs:range instead of new term belongsToClass 
- added language in prefLabel

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+  "description": "A schema for defining standard terms in openDS.",
+  "type": "object",
+  "properties": {
+    "@id": {
```
For other object we also have a @type which is the readable version and for this I presume it would be "ods:Term"? Or is there a reason this is not included

sharifX replied:
added @type similar to the others.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        "https://hdl.handle.net/20.500.1025/DigitalSpecimen"
+      ]
+    },
+    "ods:fdoType": {
```
So if I understand correctly we will get an ods term describing itself. So we will get an ods:fdoType object (https://hdl.handle.net/20.500.1025/fdoType) which has then a property which is defined by itself? Is this correct use of the terms?

sharifX replied:
Yes, this is common in RDF/ontology and data typing world. self referencing/recursive definition. 
For example in RDF schema (https://www.w3.org/2000/01/rdf-schema) 
rdfs:Class is the class of all classes, and it is itself an instance of rdfs:Class.
```
rdfs:Class a rdfs:Class ;
	rdfs:isDefinedBy <http://www.w3.org/2000/01/rdf-schema#> ;
	rdfs:label "Class" ;
	rdfs:comment "The class of classes." ;
	rdfs:subClassOf rdfs:Resource .
```

This also supports FDO Forum recommendation. (https://zenodo.org/records/14614908 page 7). 
exact quote from the pdf in the link: 
`"attribute definitions are capable to also describe definitions of attribute definitions."`

and yes, there could be performance issues related to resolution and serialisation/deserialisation. but that could be solved other ways.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        "2023-10-02T12:31:34.806Z"
+      ]
+    },
+  
```
Unnecessary whitespace

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+      ]
+    },
+  
+    "dcterms:hasVersion": {
```
How does this work together with "ods:version"

samleeflang replied:
This is still here. I still have the same question.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        ["https://terms.dissco.tech/digital-specimen-terms#EntityRelationship", "http://purl.org/dc/dcmitype/Collection"]
+      ]
+    },
+    "owl:deprecated": {
```
Shouldn't this be always present and therefore required?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        "1.0.1"
+      ]
+    },
+    "rdfs:range": {
```
Shouldn't this be always present and therefore required? When is a term not part of a class?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        "ods:DigitalSpecimen"
+      ]
+    },
+    "skos:prefLabel": {
```
Shouldn't this be always present and therefore required?

sharifX replied:
yes, added

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        "type": "string"
+      }
+    },
+    "skos:definition": {
```
Shouldn't this be always present and therefore required?

sharifX replied:
yes, added

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        1
+      ]
+    },
+ "ods:hasAgents": {
```
Shouldn't this be always present and therefore required? Shouldn't we always store a creator and a modifier?

sharifX replied:
as discussed in the meeting a few weeks ago, the term schema will have hasAgents required and call the object. we do not need to list hasRoles in term schema separately.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+        "https://doi.org/21.T11148/894b1e6cad57e921764e"
+      ]
+    },
+    "ods:status": {
```
Shouldn't this be always present and therefore required?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,148 @@
+      ]
+    }
+  },
+  "required": [
```
Missing a couple of required attributes here

samleeflang reviewed (commented):
Couple comments. General points:
- Can a term reference itself? We use odsTerms here which are described by this object. So the object is using a term which is defined by itself. Not sure if that is an issue or not.
- I think we need a lot more required attributes

samleeflang commented on data-model/fdo-type/term/examples/ods:topicDiscipline-example.json:
```diff
@@ -0,0 +1,23 @@
+  "ods:fdoType": "https://doi.org/21.T11148/bbad8c4e101e8af01115",
+  "ods:status": "ods:Active",
+  "ods:version": 1,
+  "ods:hasAgents": [],
```
This should have a creator and/or modifier

samleeflang reviewed (commented):
Have some comments, mainly about the examples.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+      "examples": [
+        "Digital Specimen"
+      ],
+      "@language": {
```
This isn't a valid JsonSchema property

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+    "skos:definition",
+    "ods:hasAgents",
+    "ods:status",  
+    "rdfs:range"
```
Would range always be there? If I got a ods:DigitalSpecimen which is the class than it won't have a range correct?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+        "type": "string"
+      }
+    },
+  "skos:example": {
```
Should we make this a required field?

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+ "ods:fdoType": {
+      "type": "string",
+      "description": "The Handle to the FDO type of the object",
+      "pattern": "^https:\/\/hdl\\.handle\\.net\/[\\w.]+\/(.){3}-(.){3}-(.){3}",
```
This is an incorrect pattern. Even the example doesn't validate against this regex.

samleeflang commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+      "description": "The Handle to the FDO type of the object",
+      "pattern": "^https:\/\/hdl\\.handle\\.net\/[\\w.]+\/(.){3}-(.){3}-(.){3}",
+      "examples": [
+        "https://hdl.handle.net/21.T11148/2ac65a933b7a0361b651"
```
This is still the VirtualCollection, better to point to the FDOType for the Term here

samleeflang reviewed (commented):
Check the comments, but there are still some basic issue with this PR.
Please validate that the properties you use to describe are valid JSON Schema properties. You can just check the standard https://json-schema.org/.
Validate the examples against the schema.
This can easily be done through something like https://www.jsonschemavalidator.net/ 
You will immediately see there are issue with the validation of the examples.

sharifX commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+        "type": "string"
+      }
+    },
+    "skos:member": {
```
ah yes, this can be removed. ods:hasControlledvalue does the same purpose justification is that the proper use of skos:member is related to skos:Collections. we are not using that. i added some notes in DEX-106.

sharifX replied:
and formatting fixed.

sharifX commented on data-model/fdo-type/term/term.json:
```diff
@@ -0,0 +1,184 @@
+        "type": "string"
+      }
+    },
+    "skos:member": {
```
this can be removed. ods:hasControlledvalue does the same purpose justification is that the proper use of skos:member is related to skos:Collections. we are not using that. i added some notes in DEX-106. and yes. each term will be a dissco vocab. ods:topicDiscipline example has this.
